### PR TITLE
Improve ARM and PPC analysis ##anal

### LINKED ISSUE #17333: Automatic analysis incorrect detects function arguments instead of no arguments
### Work environment

| Questions                                            | Answers
|------------------------------------------------------|--------------------
| OS/arch/bits (mandatory)                             | Debian x86_64 (Windows 10 Linux subsystem for Linux) [host] / EFM32LG330 arm Cortex-M3 (target)
| File format of the file you reverse (mandatory)      | ELF 32-bit
| Architecture/bits of the file (mandatory)            | ARM thumb
| r2 -v full output, **not truncated** (mandatory)     | radare2 4.6.0-git 24984 @ linux-x86-64 git.4.4.0-462-g0245b82e6 commit: 0245b82e68b0ac65a69d0724966f5ead9c29768b build: 2020-07-22__10:31:37

### Expected behavior

After loading the test binary and executing automatic analysis, the signature of the `sym.CHIP_Init` function should be:

    void sym.CHIP_Init (void);

### Actual behavior

After loading the test binary and executing automatic analysis, the signature of the `sym.CHIP_Init` function is incorrect. Analysis detected 2 arguments while the function does not take any arguments.

    void sym.CHIP_Init (int16_t arg1, int16_t arg2);

### Steps to reproduce the behavior 

    $ r2 hello-arm.axf
    > aaa
    > s sym.CHIP_Init
    > afs
    void sym.CHIP_Init (int16_t arg1, int16_t arg2);

The disassembled function does make use of `r0` (but as the return value of the `sym.SYSTEM_GetProdRev` function) and does not make use of `r1`. I don't understand why the analysis believe that the `sym.CHIP_Init` function takes two arguments.

### Additional information

```asm
            ; CALL XREF from main @ 0x2bc
┌ 64: sym.CHIP_Init (int16_t arg1, int16_t arg2);
│           ; var int16_t var_0h @ sp+0x0
│           ; arg int16_t arg1 @ r0
│           ; arg int16_t arg2 @ r1
│           0x00000270      80b5           push {r7, lr}               ; em_system.h:375
│           0x00000272      82b0           sub sp, 8
│           0x00000274      00af           add r7, sp, 0
│           0x00000276      0e4b           ldr r3, [0x000002b0]        ; em_chip.h:72 ; [0x2b0:4]=0x400c0000
│           0x00000278      0122           movs r2, 1
│           0x0000027a      1a64           str r2, [r3, 0x40]
│           0x0000027c      fff7ecff       bl sym.SYSTEM_GetProdRev    ; em_chip.h:169
│           0x00000280      0346           mov r3, r0                  ; arg1
│           0x00000282      fb71           strb r3, [r7, 7]
│           0x00000284      3b1d           adds r3, r7, 4              ; em_chip.h:170
│           0x00000286      1846           mov r0, r3                  ; int16_t arg1
│           0x00000288      fff7aaff       bl sym.SYSTEM_ChipRevisionGet
│           0x0000028c      fb79           ldrb r3, [r7, 7]            ; em_chip.h:172
│           0x0000028e      0f2b           cmp r3, 0xf
│       ┌─< 0x00000290      0ad9           bls 0x2a8
│       │   0x00000292      3b79           ldrb r3, [r7, 4]
│       │   0x00000294      022b           cmp r3, 2
│      ┌──< 0x00000296      07d9           bls 0x2a8
│      ││   0x00000298      064b           ldr r3, [0x000002b4]        ; em_chip.h:175 ; [0x2b4:4]=0x400c80c0
│      ││   0x0000029a      1b68           ldr r3, [r3]
│      ││   0x0000029c      23f05003       bic r3, r3, 0x50
│      ││   0x000002a0      044a           ldr r2, [0x000002b4]        ; em_chip.h:174 ; [0x2b4:4]=0x400c80c0
│      ││   0x000002a2      43f01003       orr r3, r3, 0x10            ; em_chip.h:175
│      ││   0x000002a6      1360           str r3, [r2]                ; em_chip.h:174
│      ││   ; CODE XREFS from sym.CHIP_Init @ 0x290, 0x296
│      └└─> 0x000002a8      00bf           nop                         ; em_chip.h:349
│           0x000002aa      0837           adds r7, 8
│           0x000002ac      bd46           mov sp, r7
└           0x000002ae      80bd           pop {r7, pc}
```

The original source code for the function:

```c
void CHIP_Init(void)
{
  MSC->CMD = MSC_CMD_INVCACHE;

  /****************************/
  /* Fix for errata CMU_E113. */

  uint8_t                     prodRev;
  SYSTEM_ChipRevision_TypeDef chipRev;

  prodRev = SYSTEM_GetProdRev();
  SYSTEM_ChipRevisionGet(&chipRev);

  if ((prodRev >= 16) && (chipRev.minor >= 3)) {
    /* This fixes an issue with the LFXO on high temperatures. */
    *(volatile uint32_t*)0x400C80C0 =
      (*(volatile uint32_t*)0x400C80C0 & ~(1 << 6) ) | (1 << 4);
  }
}
```

[hello-arm.zip](https://github.com/radareorg/radare2/files/4958874/hello-arm.zip)

### PR DESCRIPTION
**Pull request by GustavoLCR**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

* Fix stepping with ESIL into thumb code
* Run `aaef` instead of `aae` in `aaa` (this improves variable detection for functions)
* Fill src/dst info on more op types on ARM and PPC
* Fix for PPC local variables detection
* Fix arm pseudo syntax output
    * Correctly substitute pc relative offsets
    * Substitute address references to variables
* Fill op->cond for ARM64
* Fix ARM jump table size
* Implement `r_anal_block_recurse_depth_first()`
* Improve register arguments detection

**Test plan**

Tests pass

**Closing issues**

Closes #17333

### CODE REVIEW

File: test/db/cmd/cmd_pd [hunk 1]
```diff
@@ -1915,7 +1915,7 @@ EOF
 EXPECT=<<EOF
 |           0x0001045c      STR R0, [var_8h]
 |           0x00010460      STR R1, [var_ch]
-|           0x00010464      LDR R0, [envp]
```
Comment (XVilka) [inline test/db/cmd/cmd_pd:1918]: Why `envp` disappeared in the test? It is a valid argument for `main`: `int main(int argc, char *argv[], char *envp[])`:
- https://www.gnu.org/software/libc/manual/html_node/Program-Arguments.html
- https://stackoverflow.com/questions/10321435/is-char-envp-as-a-third-argument-to-main-portable

Comment (GustavoLCR) [reply test/db/cmd/cmd_pd:1918]: @XVilka If you look at the assembly `[fp, 8]` actually points to the 4th argument. `[fp, 4]` is what should point to `envp`

**Comment by XVilka** [comment]
Please resolve the conflict/rebase.

**Comment by XVilka** [comment]
```
../libr/anal/p/anal_arm_cs.c: In function ‘parse_reg_name’:
../libr/anal/p/anal_arm_cs.c:3076:20: error: ‘reg_list’ undeclared (first use in this function); did you mean ‘reg_base’?
   reg_base->name = reg_list[atoi (reg_base->name + 1)];
                    ^~~~~~~~
                    reg_base
../libr/anal/p/anal_arm_cs.c:3076:20: note: each undeclared identifier is reported only once for each function it appears in
At top level:
../libr/anal/p/anal_arm_cs.c:3085:14: warning: ‘reg_list’ defined but not used [-Wunused-variable]
 static char *reg_list[] = {
              ^~~~~~~~
```

**Comment by XVilka** [comment]
@ret2libc please take a look and merge if you are fine.

**Pull request merged**